src/main/java/org/jenkinsci/plugins/workflow/support/actions/WorkspaceRunAction.java

```diff
@@ -0,0 +1,96 @@
+/*
+ * The MIT License
+ *
+ * Copyright 2018 CloudBees, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package org.jenkinsci.plugins.workflow.support.actions;
+
+import hudson.Extension;
+import hudson.model.Action;
+import hudson.model.Item;
+import hudson.security.AccessControlled;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import jenkins.model.TransientActionFactory;
+import org.jenkinsci.plugins.workflow.flow.FlowExecutionOwner;
+import org.jenkinsci.plugins.workflow.graph.FlowNode;
+import org.jenkinsci.plugins.workflow.graphanalysis.DepthFirstScanner;
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.NoExternalUse;
+
+/**
+ * Display of all {@link WorkspaceActionImpl}s for a build.
+ */
+@Restricted(NoExternalUse.class)
+public final class WorkspaceRunAction implements Action {
+
+    public final FlowExecutionOwner owner;
+
+    WorkspaceRunAction(FlowExecutionOwner owner) {
+        this.owner = owner;
+    }
+
+    @Override public String getIconFileName() {
+        return "folder.png";
+    }
+
+    @Override public String getDisplayName() {
+        return "Workspaces";
+    }
+
+    @Override public String getUrlName() {
+        return "ws";
+    }
+
+    public List<WorkspaceActionImpl> getActions() throws Exception {
+        List<WorkspaceActionImpl> r = new ArrayList<>();
+        for (FlowNode node : new DepthFirstScanner().allNodes(owner.get())) {
```

Comment:
🐜  Can be done with new `DepthFirstScanner().filteredNodes(owner.get(), FlowScanningUtils.hasActionPredicate(WorkspaceActionImpl.class))` as a one-liner - then you iterate over that because it's an Iterable

Reply:
Might consider caching

Reply:
I actually started off using `hasActionPredicate` just like you said, but it does not really help anything since I still need to call `getAction` or similar and check for null. This is shorter and should be just as fast.

Caching is neither desirable nor necessary here: it is called once per page load, which is only when a user actually clicks the **Workspaces** link, and the result can change during a running build.

Reply:
`hasActionPredicate` does the null check for you :)

Reply:
🐛 I *think* you want to do `getOrNull()` here for UI display - so it doesn't block waiting for loading (better to just wait for a page refresh), and do a nullCheck - `get()` may cause the thread to be blocked a long time.

Reply:
I would rather have the thread blocked momentarily than potentially have no information.